app/lib/service/openid/openssl_commands.dart

```diff
@@ -0,0 +1,145 @@
+import 'package:pem/pem.dart';
+import 'package:pub_dev/shared/utils.dart';
+
+/// Randomly generated private and public RSA keypair.
+@visibleForTesting
+class RsaKeyPair {
+  final Uint8List privateKey;
+  final Uint8List publicKey;
+
+  RsaKeyPair(this.privateKey, this.publicKey);
+}
+
+/// Generates a new RSA keypair using `openssl`.
+@visibleForTesting
+Future<RsaKeyPair> generateRsaKeyPair({
+  int keySize = 2048,
+}) async {
+  return withTempDirectory((dir) async {
+    final privateKeyPath = p.join(dir.path, 'private.pem');
+    final publicKeyPath = p.join(dir.path, 'public.pem');
+    final pr1 = await runProc(
+      [
+        'openssl',
+        'genrsa',
+        '-out',
+        privateKeyPath,
+        '$keySize',
+      ],
+      timeout: Duration(seconds: 10),
+    );
+    if (pr1.exitCode != 0) {
+      throw Exception('Unable to run openssl:\n${pr1.asJoinedOutput}');
+    }
+    final privateContent = await File(privateKeyPath).readAsString();
+    final privateKeyBytes = decodePemBlocks(PemLabel.privateKey,
+            privateContent.replaceAll(' RSA PRIVATE KEY', ' PRIVATE KEY'))
+        .single;
+
+    final pr2 = await runProc(
+      [
+        'openssl',
+        'rsa',
+        '-in',
+        privateKeyPath,
+        '-pubout',
+        '-out',
+        publicKeyPath,
+      ],
+      timeout: Duration(seconds: 10),
+    );
+    if (pr2.exitCode != 0) {
+      throw Exception('Unable to run openssl:\n${pr1.asJoinedOutput}');
+    }
+    final publicContent = await File(publicKeyPath).readAsString();
+    final publicKeyBytes =
+        decodePemBlocks(PemLabel.publicKey, publicContent).single;
+    return RsaKeyPair(
+      Uint8List.fromList(privateKeyBytes),
+      Uint8List.fromList(publicKeyBytes),
+    );
+  });
+}
+
+/// Signs [input] text with the private RSA key and returns the signature bytes.
+@visibleForTesting
+Future<Uint8List> signTextWithRsa({
+  required String input,
+  required List<int> privateKey,
+}) async {
+  return await withTempDirectory((dir) async {
+    final outputFile = File(p.join(dir.path, 'output.sign'));
+    final inputFile = File(p.join(dir.path, 'input.txt'));
+    await inputFile.writeAsString(input);
+    final privateKeyFile = File(p.join(dir.path, 'private.pem'));
+    await privateKeyFile.writeAsString(
+        encodePemBlock(PemLabel.privateKey, privateKey)
+            .replaceAll(' PRIVATE KEY', ' RSA PRIVATE KEY'));
+    final pr = await runProc(
+      [
+        'openssl',
+        'dgst',
+        '-sha256',
+        '-binary',
+        '-sign',
+        privateKeyFile.path,
+        '-out',
+        outputFile.path,
+        inputFile.path,
+      ],
+      timeout: Duration(seconds: 10),
+    );
+    if (pr.exitCode != 0) {
+      throw Exception('Unable to run openssl:\n${pr.asJoinedOutput}');
+    }
+    return await outputFile.readAsBytes();
+  });
+}
+
+/// Verifies if [signature] is valid for [input], using the RSA public key.
```

Comment:
Note that `publicKey` must be a DER encoded SubjectPublicKeyInfo following RFC 5280.

I think that's the thing that this is supposed to be.

I'm deciphering this using: https://datatracker.ietf.org/doc/html/rfc7468#section-4


-----------------

I think it's generally useful to write down what bits we expect to enter a variable.